onnxruntime/core/framework/sparse_tensor.cc

```diff
@@ -3,31 +3,149 @@
-                           size_t nnz,
-                           std::shared_ptr<IAllocator> allocator)
-    : values_(elt_type, TensorShape({static_cast<int64_t>(nnz)}), allocator),
-      indices_(DataTypeImpl::GetType<int64_t>(),
-               TensorShape({static_cast<int64_t>(nnz), static_cast<int64_t>(shape.NumDimensions())}),
-               allocator),
-      shape_(shape) {}
+                           const TensorShape& dense_shape,
+                           const OrtMemoryInfo& location,
+                           std::shared_ptr<IAllocator>&& allocator)
+    : format_flags_(SparseFormatFlags::kUndefined),
+      dense_shape_(dense_shape),
+      allocator_(std::move(allocator)),
+      location_(location),
+      ml_data_type_(elt_type->AsPrimitiveDataType()),
+      rep_() {
+  ORT_ENFORCE(ml_data_type_ != nullptr, "Expecting a PrimitiveDataType as elt_type");
+}
+
+SparseTensor::SparseTensor()
+    : format_flags_(SparseFormatFlags::kUndefined),
+      dense_shape_(),
+      allocator_(),
+      location_(),
+      ml_data_type_(nullptr),
+      rep_() {
+}
+
+SparseTensor& SparseTensor::operator=(SparseTensor&& o) noexcept {
+  format_flags_ = o.format_flags_;
+  dense_shape_ = std::move(o.dense_shape_);
+  allocator_ = std::move(o.allocator_);
+  location_ = std::move(o.location_);
+  ml_data_type_ = std::move(o.ml_data_type_);
+  rep_ = std::move(o.rep_);
+  return *this;
+}
+
+SparseTensor::~SparseTensor() = default;
+
+SparseRep::SparseRep(MLDataType data_type, const TensorShape& values_shape,
+                     int64_t total_buffer_size, const AllocatorPtr& allocator)
+    : allocator_(allocator), 
+      p_data_(nullptr),
+      values_([&]() mutable -> Tensor {
+        // In the edge case when tensor is absolutely sparse, we do not have neither values, nor indices
+        if (total_buffer_size > 0) {
+          ORT_ENFORCE(values_shape.Size() * static_cast<int64_t>(data_type->Size()) < total_buffer_size,
+                      "Values size must be less than total buffer size");
+          auto data_ptr = IAllocator::MakeUniquePtr<void>(allocator_, static_cast<size_t>(total_buffer_size));
+          ORT_ENFORCE(data_ptr != nullptr, "SparseTensor Allocation failed for size: ", total_buffer_size);
+          // We own the buffer, so we must properly construct strings. Neither of the Tensors
+          // we construct on top of the buffer own it. We are constructing empty strings, hopefully
+          // nothrow and no buffer allocation
+          if (utils::IsDataTypeString(data_type)) {
+            const int64_t shape_size = values_shape.Size();
+            auto* ptr = static_cast<std::string*>(data_ptr.get());
+            for (int64_t i = 0; i < shape_size; ++i) {
+              new (ptr + i) std::string();
+            }
+          }
+          p_data_ = data_ptr.release();
+        }
+        return Tensor(data_type, values_shape, p_data_, allocator_->Info());
```

Comment:
What's the reason code in the Tensor implementation (e.g. Tensor::Init has std::string handling) can't be used for this?

Reply:
The issue here is that Tensor constructs and destructs strings only if it owns the buffer. None of the Tensors in this implementation actually own anything, they are used for projection and convinitent Tensor interface along with storing the shape.

Reply:
Could we move the std::string init logic to a shared helper instead of cut-and-pasting though?

Reply:
std::string init logic is now in the util functions and re-used across SparseTensor and Tensor classes.